### DISCUSSION TITLE
Simplify search / on_search

### DISCUSSION THREAD

**Discussion by pkapustin** [beckn/protocol-specifications#169] [Core Specification]
As long as we have registry, I suggest the following simplifications:
- Remove X-Gateway-Authorization
- BPPs send on_search directly to BAPs.

This has following benefits:
- Simpler implementation
- Better performance
- Fewer possibilities for errors
- Easier to test

**Comment by pkapustin** [comment]
I think this is in line with trying to make gateway simple and stateless: its job is just to forward all the search requests to all the providers in the domain / region.

**Comment by venkatramanm** [comment]
Yes, it would be simple. 

I can see one reason for sending to bg though. 

Bg can keep stats on performance of BPP. There may be slas that can be enforceable. In small networks, even gw is not needed for search. As one can query the registry and broadcast to bpps. 
Registry search can be a SDK function inside the bap. Basically gw and bap are really one. But unbundling gives us capability of optimizing just one place in a language agnostic way. 

--- as an aside,--
Beckn is designed around the patterns internet and dns. May be it was thought that when length of network chain is large, routing optimization would be necessary. But by that token all requests must go through gateway. Then we can keep track of transactions and keep payment ledgers etc also needed.

Seems like over design to me.

**Comment by pkapustin** [comment]
I agree that in the future the gateways may be more advanced and keep various stats, etc. For that all the requests would have to go through them. This raises all other kinds of questions, like what if BAP / BPP chooses to use a different gateway, etc.

But currently, we are not in that situation, and I think that sending on_search to the BAP along with removing X-Gateway-Authorization would simplify things a lot and allow scenarios like sending requests directly to the BPP that may be beneficial in some networks.
Currently this is not possible as both BAP and BPP have to check X-Gateway-Authorization, so the request has to pass through the gateway. 

I think at this stage of network development we would benefit from having gateway a simple and stateless entity that just forwards the requests. This way, one gateway can be substituted with another without any change in behavior.

As long as we have a registry and gateway is simply forwarding the requests, checking BAP / BPP signature should be enough, and X-Gateway-Authorization shouldn't be necessary.

**Reply by pkapustin** [reply]
@core-wg-admin What do you think?

**Comment by ravi-prakash-v** [comment]
I think it would be useful to define the purpose of a gateway and then see if it is a good-to-have or must-have entity on a network. Or describe the conditions where a BG is redundant and where it is required.

**Reply by pkapustin** [reply]
Yeah, this makes sense.

I suggest we agree that for now gateway is a good-to-have entity in the network.
Here are the reasons:

- It is only used for search/on_search.
- There are no clear functions defined for gateway other than forwarding the search request to all BPPs in that area and domain.
- There are no defined rules as to what gateway needs to be used by a BAP.
- There is no rule that the same gateway needs to be used for search and on_search.
